example/EmbeddedChessPieces/src/ECPChessLogic/ECPChessPiece.cpp

```diff
@@ -1,19 +1,149 @@
+    const int colDiff = (int) currentField.column - (int) newField.column;
+    const int rowDiff = (int) currentField.row - (int) newField.row;
+
+    if (colDiff != 0) {
+        moveHorizontally(colDiff);
+    }
+
+    if (rowDiff != 0) {
+        moveVertically(rowDiff);
+    }
 
     currentField = newField;
+    turnBackToInitialDirection();
+
     return true;
 };
 
 ECPChessField ECPChessPiece::getCurrentField() {
     return currentField;
 };
+
+void ECPChessPiece::moveHorizontally(int fieldsToMove) {
+    const Direction newDirection = fieldsToMove > 0 ? WEST : EAST;
+
+    // turn in the right direction if necessary
+    switch (currentDirection) {
+        case NORTH:
+            newDirection == WEST ? ecpMovement.turnLeft() : ecpMovement.turnRight();
+            break;
+        case EAST:
+            if (newDirection == WEST) {
+                ecpMovement.turnLeft();
+                ecpMovement.turnLeft();
+            }
+            break;
+        case SOUTH:
+            newDirection == WEST ? ecpMovement.turnRight() : ecpMovement.turnLeft();
+            break;
+        case WEST:
+            if (newDirection == EAST) {
+                // turn around
+                ecpMovement.turnLeft();
+                ecpMovement.turnLeft();
+            }
+    }
+
+    currentDirection = newDirection;
+    ecpMovement.move(abs(fieldsToMove));
+};
+
+void ECPChessPiece::moveVertically(int fieldsToMove) {
+    const Direction newDirection = fieldsToMove > 0 ? SOUTH : NORTH;
+
+    // turn in the right direction if necessary
+    switch (currentDirection) {
+        case NORTH:
+            if (newDirection == SOUTH) {
+                ecpMovement.turnLeft();
+                ecpMovement.turnLeft();
+            }
+            break;
+        case EAST:
+            newDirection == SOUTH ?  ecpMovement.turnRight() : ecpMovement.turnLeft();
+            break;
+        case SOUTH:
+            if (newDirection != SOUTH) {
+                ecpMovement.turnLeft();
+                ecpMovement.turnLeft();
+            }
+            break;
+        case WEST:
+            newDirection == SOUTH ? ecpMovement.turnLeft() : ecpMovement.turnRight();
+    }
+
+    currentDirection = newDirection;
+    ecpMovement.move(abs(fieldsToMove));
+};
+
+void ECPChessPiece::turnBackToInitialDirection() {
+    switch (currentDirection) {
+        case NORTH:
+            if (isWhite) {
+                // already facing in correct direction
+            } else {
+                ecpMovement.turnLeft();
+                ecpMovement.turnLeft();
+            }
+            break;
+        case EAST:
+            isWhite ? ecpMovement.turnLeft() : ecpMovement.turnRight();
+            break;
+        case SOUTH:
+            if (isWhite) {
+                ecpMovement.turnLeft();
+                ecpMovement.turnLeft();
+            } else {
+                // already facing in correct direction
+            }
```

Comment:
```suggestion
            }
```
I think it's understandable without the else. There is also no clarification needed in `moveVertically`.
The same for `case NORTH`, it could be a single if-statement using `!isWhite`.

Reply:
Addressed in ee3da03 and e7879f8